dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,672 @@
+        graphPopulator = new RelationGraphPopulator();
+        relModelCache = new HashMap<>();
+        relGraphCache = HashBiMap.create();
+        edgeModelCache = new HashMap<>();
+    }
+
+    public ExecutionModelManager setRelationsToExtract(List<String> relationNames) {
+        relsToExtract = relationNames;
+        return this;
+    }
+
+    public ExecutionModelNext buildExecutionModel(EncodingContext context, Model model) {
+        executionModel = new ExecutionModelNext();
+
+        this.context = context;
+        this.model = model;
+        this.wmm = context.getTask().getMemoryModel();
+        this.domain = new EventDomainNext(executionModel);
+
+        extractEvents();
+        extractMemoryLayout();
+
+        relModelCache.clear();
+        relGraphCache.clear();
+        edgeModelCache.clear();
+        extractRelations(relsToExtract);
+
+        this.context = null;
+        this.model = null;
+        this.wmm = null;
+        this.domain = null;
+
+        return executionModel;
+    }
+
+    private void extractEvents() {
+        int id = 0;
+        List<Thread> threadList = new ArrayList<>(context.getTask().getProgram().getThreads());
+
+        for (Thread t : threadList) {
+            ThreadModel tm = new ThreadModel(t);
+            int eventNum = 0;
+            Event e = t.getEntry();
+
+            do {
+                if (!isTrue(context.execution(e))) {
+                    e = e.getSuccessor();
+                    continue;
+                }
+
+                if (toExtract(e)) {
+                    extractEvent(e, tm, id++);
+                    eventNum++;
+                }
+                
+                if (e instanceof CondJump jump
+                    && isTrue(context.jumpCondition(jump))
+                ) {
+                    e = jump.getLabel();
+                } else {
+                    e = e.getSuccessor();
+                }
+
+            } while (e != null);
+
+            if (eventNum > 0) { executionModel.addThread(tm); }
+        }
+    }
+
+    private void extractEvent(Event e, ThreadModel tm, int id) {
+        EventModel em;
+        if (e.hasTag(Tag.MEMORY)) {
+            Object addressObj = checkNotNull(
+                evaluateByModel(context.address((MemoryEvent) e))
+            );
+            final BigInteger address = new BigInteger(addressObj.toString());
+            
+            String valueString = String.valueOf(
+                evaluateByModel(context.value((MemoryCoreEvent) e))
+            );
+            final BigInteger value = switch(valueString) {
+                // NULL case can happen if the solver optimized away a variable.
+                // This should only happen if the value is irrelevant, so we will just pick 0.
+                case "false", "null" -> BigInteger.ZERO;
+                case "true" -> BigInteger.ONE;
+                default -> new BigInteger(valueString);
+            };
```

Comment:
This is an oddity of the old `ExecutionModel`. For some time, we only had integer variables and then booleans were added. It was simpler to interpret them as ints to get CAAT running, but I think in general we would need something like a `Value` or `ValueModel` class that could represent the value of different types. We will require this for floating types in the future, and independent of that `BigInteger` also loses type size information (e.g. `bv32` vs. `bv64`). It might be the case that such a `ValueModel` is just a literal expression, i.e., `IntLiteral`, `BoolLiteral` etc. (not sure if this holds true for complex pointer expressions, cause those might not always fold to single literal).